File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 100001;
        
        int l = 1;
        int r = 100000;
        
        while (l <= r) {
            int mid = (l + r) / 2;
            if (isValid(diffs, times, limit, mid)) {
                answer = Math.min(answer, mid);
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return answer;
    }
    
    boolean isValid(int[] diffs, int[] times, long limit, int level) {
        long amount = times[0];
        
        for (int i = 1; i < diffs.length; i++) {
            if (diffs[i] <= level) {
                amount += times[i];
            } else {
                long timeSum = times[i] + times[i - 1];
                long level_diff = diffs[i] - level;
                amount += level_diff * timeSum;
                amount += times[i];
            }
            if (amount > limit) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
typedef long long ll;
using namespace std;

bool check(vector<int> diffs, vector<int> times, long long limit, int level) {
    ll sum = times[0];
    
    for (int i = 1; i < diffs.size(); i++) {
        int cur = diffs[i];
        if (level >= cur) {
            sum += times[i];
        } else {
            int t = cur - level;
            sum = sum + (times[i] + times[i - 1]) * t;
            sum += times[i];
        }
        if (sum > limit) return false;
    }
    return true;
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 1000001;
    
    int l = 1;
    int r = 100000;
    while (l <= r) {
       int mid = (l + r) / 2;
        if (check(diffs, times, limit, mid)) {
            answer = min(answer, mid);
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }
    return answer;
}